include/sparrow/null_layout.hpp

```diff
@@ -0,0 +1,256 @@
+// Copyright 2024 Man Group Operations Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or mplied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <cstddef>
+#include <functional>
+#include <optional>
+#include <ranges>
+
+#include "sparrow/data_type.hpp"
+#include "sparrow/iterator.hpp"
+
+namespace sparrow
+{
+
+    /*
+     * @class empty_iterator
+     *
+     * @brief Iterator used by the null_layout class.
+     *
+     * @tparam T the value_type of the iterator
+     */
+    template <class T>
+    class empty_iterator : public iterator_base<
+                               empty_iterator<T>,
+                               T,
+                               std::contiguous_iterator_tag,
+                               T>
+    {
+    public:
+
+        using self_type = empty_iterator<T>;
+        using base_type = iterator_base<
+            self_type,
+            T,
+            std::contiguous_iterator_tag,
+            T>;
+        using reference = typename base_type::reference;
+        using difference_type = typename base_type::difference_type;
+
+        explicit empty_iterator(difference_type index = difference_type()) noexcept;
+
+    private:
+
+        reference dereference() const;
```

Comment:
Here you have several warnings about using [[nodiscard]]. Check the "Files changed" panel

Reply:
Let's make a pass on this later, since it is missing for a lot of functions that were implemented before clang-format was set up.

Reply:
I've opened #121 to track it.